kubernetes/cache/kubernetes_test.go

```diff
@@ -33,3 +34,34 @@ func TestKubeGetAndListReturnKindInfo(t *testing.T) {
 		assert.Equal(kubernetes.DeploymentType, deployment.Kind)
 	}
 }
+
+// Tests that when a refresh happens, the new cache must fully load before the
+// new object is returned.
+func TestConcurrentAccessDuringRefresh(t *testing.T) {
+	require := require.New(t)
+	d := &apps_v1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "deployment", Namespace: "test",
+		},
+	}
+
+	kialiCache := newTestKialiCache(kubetest.NewFakeK8sClient(d))
+	// Prime the pump with a first Refresh.
+	kialiCache.Refresh("test")
+
+	stop := make(chan bool)
+	go func() {
+		for {
+			select {
+			case <-stop:
+				return
+			default:
+				_, err := kialiCache.GetDeployment(d.Namespace, d.Name)
+				require.NoError(err)
```

Comment:
Are you sure this branch will actually run?
And why do you need the loop?

Reply:
You can turn on debug logs for the tests and see that `GetDeployment` is being called: `make GO_TEST_FLAGS="-race -v -run=\"TestConcurrentAccessDuringRefresh\"" test`. The loop is to help ensure that we're calling `GetDeployment` at the same time that we are recreating the cache with `kialiCache.Refresh("test")`. It just calls `GetDeployment` a whole bunch of times continuously to catch race conditions where the cache is being refreshed while something is reading from the cache.

Reply:
Again, I think something confused me, but for this one I cannot remember what it was.
Perhaps, what I don't like too much is that this is kinda of doing brute forcing to catch the race conditions. So, it is non-deterministic. However, I cannot come up with a proper way of doing it.